src/test/java/com/google/neighborgood/UserInfoServletTest.java

```diff
@@ -0,0 +1,422 @@
+    }
+
+    // The POST request should lead to an error handling clause, there should be no entity in the
+    // datastore
+    assertEquals(0, ds.prepare(new Query("UserInfo")).countEntities(withLimit(10)));
+
+    // Now we give the servlet an invalid phone number input
+    when(request.getParameter("address-input")).thenReturn("4xxx Centre Avenue");
+    when(request.getParameter("phone-input")).thenReturn("");
+
+    try {
+      new UserInfoServlet().doPost(request, response);
+    } catch (IOException e) {
+      assertTrue(false);
+    }
+    assertEquals(0, ds.prepare(new Query("UserInfo")).countEntities(withLimit(10)));
+
+    // Finally we give the servlet a valid phone number input
+    when(request.getParameter("phone-input")).thenReturn("4xxxxxxxxx");
+
+    try {
+      new UserInfoServlet().doPost(request, response);
+    } catch (IOException e) {
+      assertTrue(false);
+    }
+
+    assertEquals(1, ds.prepare(new Query("UserInfo")).countEntities(withLimit(10)));
+    PreparedQuery results = ds.prepare(new Query("UserInfo"));
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      // The entity can't be null
+      assertTrue(false);
+    }
+
+    // Test the stored personal information
+    assertEquals("Leonard", (String) entity.getProperty("nickname"));
+    assertEquals("4xxx Centre Avenue", (String) entity.getProperty("address"));
+    assertEquals("4xxxxxxxxx", (String) entity.getProperty("phone"));
+    assertEquals("leonardzhang@google.com", (String) entity.getProperty("email"));
+    assertEquals("1234567890", (String) entity.getProperty("userId"));
+    assertEquals(0, (long) entity.getProperty("points"));
+  }
+
+  @Test
+  public void doGetTest() throws IOException {
+    HttpServletRequest request = mock(HttpServletRequest.class);
+    HttpServletResponse response = mock(HttpServletResponse.class);
+
+    // Check whether the datastore is empty before the test
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    assertEquals(0, ds.prepare(new Query("UserInfo")).countEntities(withLimit(10)));
+
+    Entity dummy = new Entity("UserInfo");
+    dummy.setProperty("nickname", "Leonard");
+    dummy.setProperty("address", "xxx");
+    dummy.setProperty("phone", "xxx");
+    dummy.setProperty("email", "test@example.com");
+    dummy.setProperty("userId", "1234567890");
+    dummy.setProperty("points", 0);
+    ds.put(dummy);
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    try {
+      new UserInfoServlet().doGet(request, response);
+    } catch (IOException e) {
+      assertTrue(false);
+    }
+
+    // After sending the GET request, the doGet function should output the json string
+    writer.flush();
+    assertTrue(stringWriter.toString().contains("[\"Leonard\",\"xxx\",\"xxx\"]"));
+  }
+
+  @Test
+  public void stressTest() {
```

Comment:
This one fails for me sometimes (it's flaky). More about that below.

I see this error:

```
-------------------------------------------------------------------------------
Test set: com.google.neighborgood.servlets.UserInfoServletTest
-------------------------------------------------------------------------------
Tests run: 6, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 4.122 sec <<< FAILURE! - in com.google.neighborgood.servlets.UserInfoServletTest
stressTest(com.google.neighborgood.servlets.UserInfoServletTest)  Time elapsed: 3.457 sec  <<< FAILURE!
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.google.neighborgood.servlets.UserInfoServletTest.stressTest(UserInfoServletTest.java:407)
```

Reply:
This flaky behavior is caused by the stressTest, due to the running time of a search is not stable. As discussed yesterday, I decided to delete the stress test for now, and consider to add it back after I find a better, more stable way to carry out stress test.